provider/provider.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	// Store both outputs and inputs into the state.
+	checkpoint, err := plugin.MarshalProperties(
+		checkpointObject(inputs, outputs.Mappable()),
+		plugin.MarshalOptions{
+			Label:        fmt.Sprintf("%s.checkpoint", label),
+			KeepSecrets:  true,
+			KeepUnknowns: true,
+			SkipNulls:    true,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rpc.CreateResponse{
+		Id:         id,
+		Properties: checkpoint,
+	}, nil
+}
+
+// Read the current live state associated with a resource.
+func (p *dockerNativeProvider) Read(ctx context.Context, req *rpc.ReadRequest) (*rpc.ReadResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.Read(%s)", p.name, urn)
+	logging.V(9).Infof("%s executing", label)
+	id := req.GetId()
+
+	// For now, we will return the minimum implementation for Read,
+	// until we find a use case for reading an image from a registry and comparing it to existing state
+	return &rpc.ReadResponse{Id: id}, nil
+
+}
+
+// Update updates an existing resource with new values.
+func (p *dockerNativeProvider) Update(ctx context.Context, req *rpc.UpdateRequest) (*rpc.UpdateResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.Update(%s)", p.name, urn)
+	logging.V(9).Infof("%s executing", label)
+
+	// Read the inputs to persist them into state.
+	newInputs, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.newInputs", label),
+		KeepUnknowns: true,
+		RejectAssets: true,
+		KeepSecrets:  true,
+		SkipNulls:    true,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "diff failed because malformed resource inputs")
+	}
+	// When the docker image is updated, we build and push again.
+	_, outputProperties, err := p.dockerBuild(ctx, urn, req.GetNews())
+	if err != nil {
+		return nil, err
+	}
+	outputs, err := plugin.UnmarshalProperties(outputProperties, plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.outputs", label),
+		KeepUnknowns: true,
+		RejectAssets: true,
+		KeepSecrets:  true,
+		SkipNulls:    true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Store both outputs and inputs into the state and return RPC checkpoint.
+	checkpoint, err := plugin.MarshalProperties(
+		checkpointObject(newInputs, outputs.Mappable()),
+		plugin.MarshalOptions{
+			Label:        fmt.Sprintf("%s.checkpoint", label),
+			KeepSecrets:  true,
+			KeepUnknowns: true,
+			SkipNulls:    true,
+		},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+	return &rpc.UpdateResponse{
+		Properties: checkpoint,
+	}, nil
+}
+
+// Delete tears down an existing resource with the given ID.  If it fails, the resource is assumed
+// to still exist.
+func (p *dockerNativeProvider) Delete(ctx context.Context, req *rpc.DeleteRequest) (*pbempty.Empty, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.Update(%s)", p.name, urn)
+	logging.V(9).Infof("%s executing", label)
+	// Implement Delete logic specific to your provider.
```

Comment:
Is that going to be impl later or there is no good way to delete?

Reply:
If we need to delete, we can implement later. Vivek and I talked about some potential use cases. In the meantime, I'll remove some of this boilerplate cruft.